File: Extra Credit/Bonus Questions: "Everything Everywhere All At Once"/implementations.py
class ExtraCredit:
    @staticmethod
    def adjacency_list_to_matrix(adj_list):
        n = len(adj_list)
        # Create an n x n matrix initialized with 0s
        adj_matrix = [[0] * n for _ in range(n)]
        
        for i, neighbors in enumerate(adj_list):
            for neighbor in neighbors:
                adj_matrix[i][neighbor] = 1
        
        return adj_matrix

    @staticmethod
    def adjacency_matrix_to_list(adj_matrix):
        adj_list = []
        for i in range(len(adj_matrix)):
            # Find all indices of columns with a 1 (edge) in row i
            neighbors = [j for j in range(len(adj_matrix[i])) if adj_matrix[i][j] == 1]
            adj_list.append(neighbors)
        
        return adj_list

    @staticmethod
    def reverse_graph(adj_list):
        n = len(adj_list)
        reversed_list = [[] for _ in range(n)]
        
        for i, neighbors in enumerate(adj_list):
            for neighbor in neighbors:
                reversed_list[neighbor].append(i)
        
        return reversed_list

# Example usage:
# adj_list = [[1, 2], [0, 2], [0, 1]]
# adj_matrix = ExtraCredit.adjacency_list_to_matrix(adj_list)
# reversed_list = ExtraCredit.reverse_graph(adj_list)

File: Extra Credit/Bonus Questions: "Everything Everywhere All At Once"/implementations.java
public class ExtraCredit {
    
    public static int[][] adjacencyListToMatrix(List<List<Integer>> adjList) {
        int n = adjList.size(); // Number of vertices
        int[][] adjMatrix = new int[n][n]; // Initialize n x n matrix with 0s
        
        for (int i = 0; i < n; i++) {
            for (int neighbor : adjList.get(i)) {
                adjMatrix[i][neighbor] = 1; // Set to 1 where there's an edge
            }
        }
        
        return adjMatrix;
    }

    public static List<List<Integer>> adjacencyMatrixToList(int[][] adjMatrix) {
        List<List<Integer>> adjList = new ArrayList<>();
        int n = adjMatrix.length;
        
        for (int i = 0; i < n; i++) {
            List<Integer> neighbors = new ArrayList<>();
            for (int j = 0; j < n; j++) {
                if (adjMatrix[i][j] == 1) {
                    neighbors.add(j);
                }
            }
            adjList.add(neighbors);
        }
        
        return adjList;
    }

    public static List<List<Integer>> reverseGraph(List<List<Integer>> adjList) {
        int n = adjList.size(); // Number of vertices
        List<List<Integer>> reversedList = new ArrayList<>(Collections.nCopies(n, new ArrayList<>()));

        for (int i = 0; i < n; i++) {
            for (int neighbor : adjList.get(i)) {
                reversedList.get(neighbor).add(i); // Add i to the adjacency list of its neighbor
            }
        }
        
        return reversedList;
    }
    
}

